### PR TITLE
[Mellanox] Update headroom calculation algorithm

### PR DESCRIPTION
**Pull request by stephenxs**
**What I did**

1. Fix a bug in lossless headroom calculation regarding the small packet percentage which should be calculated in a size-based approach instead of packet-based
2. Adjust Mellanox-specific lossless headroom calculation algorithm

**Why I did it**

**How I verified it**

Manual test
Regression test https://github.com/sonic-net/sonic-mgmt/pull/14229

**Details if related**

### CODE REVIEW

**Comment by stephenxs** [comment]
The manual test result is as below. Deviation is small and acceptable
port MTU: 9100
RDMA MTU: 1024
Small packet percentage: 50%
SPC3
- 400G, 300m.
  - by lua: xon 38912, xoff 272384, size 320512
  - by using internal tool: xon 38k, xoff 265.76k, size 312.8k
- 100G, 40m.
  - by lua: xon 19456, xoff 51200, size 70656
  - by using internal tool: xon 19k, xoff 49.62k, size 68.6k

SPC4
- 800G, 300m.
  - by lua: xon 38912 xoff 474112 size 522240
  - by using internal tool: xon 38k xoff 463.01 size 510.0
- 200G, 40m
  - by lua: xon 19456 xoff 65536 size 84992
  - by using internal tool: xon 19k xoff 63.88k size 84.9k

**Review by keboliu — APPROVED** [review]

File: cfgmgr/buffer_headroom_mellanox.lua [hunk 1]
```diff
@@ -123,7 +128,7 @@ local speed_overhead
 
 -- Adjustment for 8-lane port
 if is_8lane ~= nil and is_8lane then
-    pipeline_latency = pipeline_latency * 2 - 1024
+    pipeline_latency = pipeline_latency * 2
```
Comment (kperumalbfn) [inline cfgmgr/buffer_headroom_mellanox.lua:131]: Why double the pipeline latency for 8lane?

Comment (stephenxs) [reply cfgmgr/buffer_headroom_mellanox.lua:131]: Hi
This is according to hardware implementation.

File: cfgmgr/buffer_headroom_mellanox.lua [hunk 2]
```diff
@@ -79,6 +79,11 @@ for i = 1, #asic_table_content, 2 do
     end
 end
 
+local kb_on_tile = 0
+if asic_keys[1]:sub(-1) == '4' or asic_keys[1]:sub(-1) == '5' then
+    kb_on_tile = port_speed / 1000 * 120 / 8
```
[the remark this answers is not in the document]
Comment (stephenxs) [reply cfgmgr/buffer_headroom_mellanox.lua:84]: No. It's a fixed number to calculate this parameter.

**Comment by kperumalbfn** [comment]
LGTM

**Review by kperumalbfn — APPROVED** [review]

**Review by prsunny — APPROVED** [review]

**Comment by bingwang-ms** [comment]
Cherry-pick was done by #3342

**Comment by zhangyanzhao** [comment]
proposed for 202411 release but cherry pick to 202405

**Pull request merged**